Parser seperation

MengnanLi91 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
This PR is designed to separate current MOOSE parser into a front part and end part. The front part will parse before MooseApp is built. 

## Reason
<!--Why do you need this feature or what is the enhancement?-->
This work is part of the NEAMS task: Application block to host: dynamic loading params

## Design
<!--A concise description (design) of the enhancement.-->
The plan is to separate the part of parser which is not related to MooseApp as front parser and keep the rest as the new parser.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
This will help the new capability Application block to host

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

MengnanLi91 committed:
Initial commit on parser seperation(in progress) #25994

MengnanLi91 committed:
Parser separation and add parameter duplication check before file merge

MengnanLi91 committed:
Add parameter duplication test for multiple input files merge

MengnanLi91 committed:
Update main.c file after parser change, change parser to parserOther in some calls

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1887408/) on 6d26a33 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25994/site)

This comment will be updated on new commits.

MengnanLi91 committed:
Use MooseCreate in UnitTest, fix most test cases

MengnanLi91 committed:
Fix MeshDivisionName error

MengnanLi91 committed:
Change Parserother to Builder, clean up Main.C and move to mooseMain.C, add docs

Co-Authored-By: Cody Permann <3036960+permcody@users.noreply.github.com>
Co-Authored-By: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1911769/) on d9cef40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6a86072ce25d3d6e2924bdeeec91c66a5a7833c0`

MengnanLi91 committed:
Clean format for precheck

Add missing header files

Add MooseApp header in MooseMain.h

Use new main.C in ex13 and ex15

MengnanLi91 committed:
Add missing header, update test file

Remove  #pragma once

MengnanLi91 committed:
Add alternate function to help MOOSE application adapting to main function changes in framework

Clean format for precheck

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1914825/) on 02d49bf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 37692f52f0738e005683d1745eac37b3f127e505`

MengnanLi91 committed:
move warning message to function and use mooseDeprecated

Modify warning message

MengnanLi91 committed:
Update createAppShared in module unit test

MengnanLi91 committed:
Add Helper functions to bypass new parser changes for Apps

Modify helper functions

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1920730/) on db5406e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8cad10d94553be088dd177ac8edd3833a542c3b4`

MengnanLi91 commented:
@loganharbour I've done all the changes I could at this point. Marmot and Griffin failure will need to be patched on app side given the error message(I will talk to Patrick about all the app patches after this PR has been merged) and I need some help with MooseServerTest failure. Could you review this PR when you have time?

loganharbour commented on examples/ex01_inputfile/src/main.C:
```diff
@@ -16,9 +16,8 @@
+#include "MooseMain.h"
 
 // Create a performance log
 PerfLog Moose::perf_log("Example");
```
@roystgnr can we move this around?

MengnanLi91 replied:
I moved `perf_log()` into `main()`, it compiles fine but failed all the tests as below. Does this a registration error?
`symbol not found in flat namespace '__ZN5Moose8perf_logE'`
Here is where I moved `perf_log()`, did I do anything wrong?
```
template <typename DefaultAppType>
void
main(int argc, char * argv[])
{
  MooseInit init(argc, argv);

  DefaultAppType::registerApps();

  const auto default_app_name = MooseUtils::prettyCppType<DefaultAppType>();

  // Create a performance log
  PerfLog perf_log(default_app_name);

  auto app = createMooseApp(default_app_name, argc, argv);

  app->run();
}
```

roystgnr replied:
Sorry, missed the notification somehow.

That linker error looks like the definition of `perf_log` has been moved but there's still a *declaration* of `perf_log` (possibly one being used!) which hasn't.

loganharbour commented on framework/include/parser/Builder.h:
```diff
@@ -0,0 +1,286 @@
+  Builder & _builder;
+};
+
+/**
+ * Class for parsing input files. This class utilizes the GetPot library for actually tokenizing and
+ * parsing files. It is not currently designed for extensibility. If you wish to build your own
+ * parser, please contact the MOOSE team for guidance.
+ */
```
```suggestion
/**
 * Parses MOOSE input using HIT/WASP.
 */
```

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -350,6 +351,12 @@ class MooseApp : public ConsoleStreamInterface,
+   * Helper function to link new added Builder back to Parser. This function will be removed after
+   *new Parser and builder are merged
+   **/
+  Builder & parser() { return _builder; }
```
Can we move this to the source file and add a `mooseDeprecated()` message? Something like

> MooseApp::parser() is deprecated, use MooseApp::builder() instead.

loganharbour commented on framework/src/base/AppFactory.C:
```diff
@@ -54,6 +58,58 @@ AppFactory::createAppShared(const std::string & default_app_type,
+                            char ** argv,
+                            MPI_Comm comm_world_in)
+{
+  mooseDeprecated("Please update your main.C to adapt new main function in MOOSE framework, see "
+                  "the example of ../moose/test/src/main.C. ");
```
```suggestion
  mooseDeprecated("Please update your main.C to adapt new main function in MOOSE framework, see "
                  "test/src/main.C in MOOSE as an example of moose::main() ");
```

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -959,10 +961,16 @@ MooseApp::setupOptions()
     if (!isParamValid("_input_text"))
-      _parser.parse(_input_filenames);
+    {
+      if (_input_filenames.empty())
+        mooseError("No input files specified. Add -i <inputfile> to your command line.");
```
Did this error not exist before?

MengnanLi91 replied:
This error used to be in Parser.C
```
if (_input_filenames.empty())
    mooseError("No input files specified.");
```
I moved it here because we want to allow empty input file or wrong input file type(e.g. input.e) in front parser and deal with all these options later in MooseApp.C, otherwise we will need add a lot of parameters like 
`"--show-outputs"`, `"--mesh-only [mesh_file_name]"`,` "--definition" `, etc. to new Parser along with the condition logic

loganharbour commented on framework/src/base/MooseMain.C:
```diff
@@ -0,0 +1,50 @@
+#include <omp.h>
+#endif
+
+namespace moose
```
I think the common namespace is `Moose`, not `moose`

loganharbour commented on framework/src/base/MooseServer.C:
```diff
@@ -266,7 +271,8 @@ MooseServer::getAllValidParameters(InputParameters & valid_params,
 {
   // gather global parameters then action parameters then object parameters
-  valid_params = Parser::validParams();
+  valid_params = Builder::validParams();
+  valid_params += Builder::validParams();
```
Why `=` and `+=` ?

MengnanLi91 replied:
Good catch!

loganharbour commented on test/tests/misc/app_name/tests:
```diff
@@ -3,7 +3,7 @@
     cli_args = '--type'
     expect_out = "MooseTestApp"
-    input = 'IGNORED'
+    input = ''
```
Why all of the changes from

```
input = 'IGNORED'
```

to

```
input = ''
```

?

MengnanLi91 replied:
This is suggested by @permcody.  In MooseApp.C we have a very long `if elseif and else` condition to deal with different command line options. When we want to ignore input files, however we actually don't want to use 'IGNORED' as a key word to detect "ignore input file condition", it just fells into the else condition in the very end. So we decide to replace the 'IGNORED' key word in test with ''. Therefore the logic below can work, otherwise it will looking for input file named 'IGNORED' in the folder.
```
  input_param.addCommandLineParam<std::vector<std::string>>(
      "input_file",
      "-i <input_files>",
      "Specify one or multiple input files. Multiple files get merged into a single simulation "
      "input.");

  command_line->addCommandLineOptionsFromParams(input_param);

  std::vector<std::string> input_filename;
  command_line->search("input_file", input_filename);

  if (!input_filename.empty())
    front_parser->parse(input_filename);
```

loganharbour replied:
Works for me.

loganharbour commented on test/tests/parser/param_duplicate_check/tests:
```diff
@@ -0,0 +1,10 @@
+[Tests]
+  design = 'Parser.md'
+  issues = '#25994'
+  [dup_check]
```
What new capability was added to require this test? OR is this just adding coverage that didn't exist before?

MengnanLi91 replied:
This is more like a patch to the existing duplication check. Joshua found the duplication check in parser doesn't work properly when multiple input files are merged together(only the last input file get checked). I changed the sequence of duplication and file merge to fix it. Probably should start a new PR on this, but it's very simple so I throw it here.

MengnanLi91 committed:
Address review: use Moose namespace, improve doco, add deprecated message

Co-Authored-By: Logan Harbour <loganhharbour@gmail.com>

Use Moose namespace instead of moose in modules

MengnanLi91 committed:
Use Moose namespace in modules and tutorials

loganharbour commented:
Looks like there's a few things still sticking around. did you fix and just not resolve?

MengnanLi91 commented:
> Looks like there's a few things still sticking around. did you fix and just not resolve?

Yes, I have fixed those but not sure if I should resolve before you take a look

brandonlangley commented:
@MengnanLi91 @loganharbour -

I saw this and was curious about the `MooseServerTest` seg-fault failure so I took a look into what might be happening.

I found that there are two issues with how the `MooseServer` is interacting with this `Parser` / `Builder` split right now.

There are likely other ways to fix these, but I wanted to share my findings and minimal change fixes in case it's helpful.

- - - - -

**Issue <span>#</span>1:**

  - There is an expectation here that the `Parser` already parsed the input at the time the `Builder` gets constructed.

    This is because `Builder` sets its private `_root` member using `_root(parser.getRootNode())` in the constructor.

    Then the `_root` never changes so it is invalid if the `Parser` will parse the input after the `Builder` is constructed.

    This invalid `_root` is later used throughout `Builder` processing and accessed by the `MooseServer` downstream.

  - All these other cases work since their `Parser` does parse the input prior to `MooseApp` and `Builder` construction:

     - `framework/src/multiapps/MultiApp.C` - `createApp()`

     - `framework/src/base/MooseMain.C` - `createMooseApp()`

     - `framework/src/base/AppFactory.C` - `createAppShared()`

    But `MooseServer` does not do `front_parser->parse()` before the `AppFactory::instance().createShared()` call.

**Fix for <span>#</span>1:**

  - Add a `setRootNode()` method to `Builder` that allows its `_root` member to be updated after its construction like:

    ```cpp
    void setRootNode(hit::Node * root) { _root = root; }
    ```

    And call it in the `MooseServer` case in `MooseApp` to set the `_builder` root after the parse but before the build like:

    ```cpp
    _parser->parse(_input_filenames, getParam<std::string>("_input_text"));
    _builder.setRootNode(_parser->getRootNode());
    _builder.builder();
    ```

    Then the `_root` member of `_builder` will be valid at the necessary time of processing and `MooseServer` access.

- - - - -

**Issue <span>#</span>2:**

  - The `MooseServer` got updated here so that it accesses the `_root` member of `Builder` rather than `Parser` with:

    ```cpp
    _check_app->builder().root()->getNodeView()
    ```

    Separating the input processing into sequential stages leads to a problematic scenario with this access to the root.

    If the first `Parser` stage fails due to an input check error, then the `_root` member of the `Builder` never gets set.

    So `MooseServer` cannot use the `_root` member of `Builder` if any input error is reported from the `Parser` stage.

    This currently prevents features such as autocompletion and hierarchal tree creation from working for these cases.

**Fix for <span>#</span>2:**

  - Update the `Builder::root()` method so that it returns `_parser.getRootNode()` instead its own `_root` member.

    Note `Parser::getRootNode()` currently calls `_root.release()` which releases this root from `Parser` ownership.

    Change the `getRootNode()` return from `_root.release()` to `_root.get()` so the `Parser` keeps the ownership.

    This works because `Builder` already owns a `Parser` reference keeping its `_root` in scope for the `Builder` life.

    The `Builder` destructor should also get updated to be empty since will not be handling the `_root` deletion now.

- - - - -

Here is my whole `git diff` change set, and other paths may also fix the issues but hopefully the diagnosis is useful.

```diff
diff --git a/framework/include/parser/Builder.h b/framework/include/parser/Builder.h
index 996f3c0906..4e57f688ee 100644
--- a/framework/include/parser/Builder.h
+++ b/framework/include/parser/Builder.h
@@ -114,7 +114,9 @@ public:
    */
   std::string hitCLIFilter(std::string appname, const std::vector<std::string> & argv);
 
-  hit::Node * root() { return _root; };
+  hit::Node * root() { return _parser.getRootNode(); };
+
+  void setRootNode(hit::Node * root) { _root = root; }
 
 protected:
   /**
diff --git a/framework/include/parser/Parser.h b/framework/include/parser/Parser.h
index aaa05c170f..5b32ead643 100644
--- a/framework/include/parser/Parser.h
+++ b/framework/include/parser/Parser.h
@@ -123,7 +123,7 @@ public:
   {
     // normal input file is provided, front parser is set to parse input file
     if (_root)
-      return _root.release();
+      return _root.get();
     // no valid input file is provided, front parser is skipped
     else
       return nullptr;
diff --git a/framework/src/base/MooseApp.C b/framework/src/base/MooseApp.C
index 407f315403..ffb983289b 100644
--- a/framework/src/base/MooseApp.C
+++ b/framework/src/base/MooseApp.C
@@ -969,6 +969,7 @@ MooseApp::setupOptions()
     else
     {
       _parser->parse(_input_filenames, getParam<std::string>("_input_text"));
+      _builder.setRootNode(_parser->getRootNode());
       _builder.builder();
     }
     if (isParamValid("mesh_only"))
diff --git a/framework/src/parser/Builder.C b/framework/src/parser/Builder.C
index 4f128f2578..528bb70556 100644
--- a/framework/src/parser/Builder.C
+++ b/framework/src/parser/Builder.C
@@ -134,7 +134,7 @@ Builder::Builder(MooseApp & app, ActionWarehouse & action_wh, Parser & parser)
   _input_filenames = _parser.getInputFileNames();
 }
 
-Builder::~Builder() { delete _root; }
+Builder::~Builder() {}
 
 InputParameters
 Builder::validParams()
```

loganharbour commented:
@brandonlangley the idea of having `Builder::root()` only sometimes return the root that is set seems dangerous and like a poor design

MengnanLi91 commented:
@brandonlangley Thanks a lot for the suggestions! I've adapted the changes but it still failed MooseServerTest.DocumentChangeAndDiagnostics

brandonlangley commented:
@MengnanLi91

>  I've adapted the changes but it still failed MooseServerTest.DocumentChangeAndDiagnostics

That's because you removed the trailing space from these two diagnostic messages in `MooseServerTest.C` on this PR.

Your editor likely did this automatically. Those two space characters are annoying and I'm going to fix that in another PR.

```diff
TEST_F(MooseServerTest, DocumentChangeAndDiagnostics)
{
  . . .
  std::string diagnostics_expect = R"INPUT(
line:18 column:0 - (BCs/all/boundary):
line:18 column:0 -     the following side sets (ids) do not exist on the mesh: top (2), bottom (3)
- line:18 column:0 - MOOSE distinguishes between "node sets" and "side sets" depending on whether 
- line:18 column:0 - you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding 
+ line:18 column:0 - MOOSE distinguishes between "node sets" and "side sets" depending on whether
+ line:18 column:0 - you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
line:18 column:0 -     to your side sets are constructed for you by default.
line:18 column:0 -     Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
line:18 column:0 -     Note: If you are running with adaptivity you should prefer using side sets.
)INPUT";
  EXPECT_EQ(diagnostics_expect, "\n" + diagnostics_actual.str());
}
```

brandonlangley commented:
@loganharbour

> the idea of having Builder::root() only sometimes return the root that is set seems dangerous and like a poor design

I'm not sure I understand. That is exactly what is happening with the current design before the update from my [comment](https://github.com/idaholab/moose/pull/25994#issuecomment-1853234892).

```cpp
hit::Node * Builder::root() { return _root; };
```

Thats why, as described in the **Issue <span>#</span>2** part of my comment, this call had been failing with a seg-fault as seen in testing.

```cpp
_check_app->builder().root()->getNodeView()
```

Because in any case where the `Parser` first stage fails due an input check,  the `_root `member of `Builder` is never set.

So currently, `Builder::root()` only sometimes returns the root.

Unless I'm misunderstanding something, the update that I used actually fixes the call so it will **_always_** return the root with:

```cpp
hit::Node * root() { return _parser.getRootNode(); };
```

And as I mentioned in my comment, there definitely could be some other approach to fixing this which is more preferred.

I mostly wanted to help in understanding what is happening and thought I'd also share how I fixed it in case it was useful.

loganharbour commented:
Sorry @brandonlangley, I should have clarified. This is what confuses me:

```
hit::Node * root() { return _parser.getRootNode(); };
void setRootNode(hit::Node * root) { _root = root; }
```

I don't understand why both of these methods would be necessary. If `root()` is always correct, I don't follow why we would want to also have `setRootNode()`. Having `_root != root()` is what is concerning to me

brandonlangley commented:
> Sorry @brandonlangley, I should have clarified. This is what confuses me:
> 
> ```
> hit::Node * root() { return _parser.getRootNode(); };
> void setRootNode(hit::Node * root) { _root = root; }
> ```
> 
> I don't understand why both of these methods would be necessary. If `root()` is always correct, I don't follow why we would want to also have `setRootNode()`. Having `_root != root()` is what is concerning to me

@loganharbour -

Yes we're on the same page, that makes complete sense and I also don't think both of those methods should be necessary.

I actually don't think the `Builder` class needs its own private `_root` member at all, therefore `setRootNode()` isn't needed.

The "true" `_root` is the `std::unique_ptr<hit::Node>` in the `Parser` which is why I suggested for the `Builder` call to be:

```
hit::Node * Builder::root() { return _parser.getRootNode(); };
```

while also letting `Parser` keep `_root` ownership and return `_root.get()` instead of `_root.release()` in `getRootNode()`.

The update that I shared was just the most minimal change set that fixed the issues in order to demonstrate what they were.

I too thought about removing the `_root` member from the `Builder` which removes the need for `Builder::setRootNode()`.

And then just let all usages of `_root` in the `Builder` processing logic go through that `Builder::root()` interface instead.

But this would have required updating like 70 references in `Builder` to `_root` and the change set might be less digestible.

So like I said, I figured my five-line fix would help demonstrate the issue and then you guys could fix it however you wanted.

But yeah, we're in agreement that `setRootNode()` should not be needed if `Builder` uses `root()` and not `_root` directly.

MengnanLi91 committed:
Let Parser keep the ownership of _root, use root() in Builder, trailing space in MooseServerTest

Co-Authored-By: Logan Harbour <loganhharbour@gmail.com>
Co-Authored-By: brandonlangley <21372118+brandonlangley@users.noreply.github.com>

MengnanLi91 committed:
Add MooseMain.h header file for module unit tests

brandonlangley commented:
@MengnanLi91 @loganharbour -

Three other minor suggestions came to mind as I looked through this, so I thought I would share but none are necessary.

1. Both `Parser.h` and `Builder.h` can remove the `friend class MooseServer` lines since those are no longer needed.

2. Rename `Parser::getRootNode()` to `Parser::root()` for consistency with `Builder::root()` as the use is the same.

3. Rename `Builder::builder()` (that runs the build) to `Builder::build()` as it is an action like the `parse()` method.

    - prevents mixing up with the `Builder::Builder()` constructor or `MooseApp::builder()` that returns a reference

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1947473/) on 6e5a773 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.06%</td>
      <td>85.05%</td>
      <td>-0.01%</td>
      <td>87.26%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>97860</td>
      <td>97915</td>
      <td>+55</td>
      <td>1082</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17186</td>
      <td>17212</td>
      <td>+26</td>
      <td>158</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/framework)


## Modules coverage

### Chemical reactions

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.54%</td>
      <td>92.53%</td>
      <td>-0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1440</td>
      <td>1436</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>116</td>
      <td>116</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/chemical_reactions)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/chemical_reactions)

### Combined

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>73.77%</td>
      <td>71.93%</td>
      <td>-1.84%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>45</td>
      <td>41</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>16</td>
      <td>16</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/combined)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/combined)

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.17%</td>
      <td>90.16%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4007</td>
      <td>4003</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>437</td>
      <td>437</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/contact)

### Electromagnetics

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>94.31%</td>
      <td>94.27%</td>
      <td>-0.03%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>497</td>
      <td>494</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>30</td>
      <td>30</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/electromagnetics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/electromagnetics)

### External petsc solver

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.37%</td>
      <td>-0.11%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>353</td>
      <td>350</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>60</td>
      <td>60</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/external_petsc_solver)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/external_petsc_solver)

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.24%</td>
      <td>85.23%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7194</td>
      <td>7190</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1246</td>
      <td>1246</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/fluid_properties)

### Fsi

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.00%</td>
      <td>87.90%</td>
      <td>-0.10%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>330</td>
      <td>327</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>45</td>
      <td>45</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/fsi)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/fsi)

### Functional expansion tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.76%</td>
      <td>81.72%</td>
      <td>-0.05%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>946</td>
      <td>943</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>211</td>
      <td>211</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/functional_expansion_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/functional_expansion_tools)

### Geochemistry

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.18%</td>
      <td>96.17%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>5157</td>
      <td>5154</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>205</td>
      <td>205</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/geochemistry)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/geochemistry)

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.28%</td>
      <td>88.27%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3932</td>
      <td>3928</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>522</td>
      <td>522</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/heat_transfer)

### Level set

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.53%</td>
      <td>85.41%</td>
      <td>-0.12%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>325</td>
      <td>322</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>55</td>
      <td>55</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/level_set)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/level_set)

### Misc

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>33.65%</td>
      <td>32.37%</td>
      <td>-1.28%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>71</td>
      <td>67</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>140</td>
      <td>140</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/misc)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/misc)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.23%</td>
      <td>84.22%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13194</td>
      <td>13190</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2471</td>
      <td>2471</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/navier_stokes)

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.92%</td>
      <td>88.90%</td>
      <td>-0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1781</td>
      <td>1778</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>222</td>
      <td>222</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/optimization)

### Peridynamics

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>77.50%</td>
      <td>77.47%</td>
      <td>-0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3072</td>
      <td>3068</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>892</td>
      <td>892</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/peridynamics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/peridynamics)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.63%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13140</td>
      <td>13136</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2205</td>
      <td>2205</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/phase_field)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.32%</td>
      <td>95.32%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11132</td>
      <td>11129</td>
      <td>-3</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>546</td>
      <td>546</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/porous_flow)

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.27%</td>
      <td>95.27%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4315</td>
      <td>4312</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>214</td>
      <td>214</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/ray_tracing)

### Rdg

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>63.42%</td>
      <td>63.27%</td>
      <td>-0.15%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>456</td>
      <td>453</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>263</td>
      <td>263</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/rdg)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/rdg)

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.22%</td>
      <td>95.22%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>5137</td>
      <td>5134</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>258</td>
      <td>258</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/reactor)

### Richards

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.12%</td>
      <td>93.11%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3327</td>
      <td>3323</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>246</td>
      <td>246</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/richards)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/richards)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.40%</td>
      <td>91.26%</td>
      <td>-0.14%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>170</td>
      <td>167</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>16</td>
      <td>16</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/scalar_transport)

### Solid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.77%</td>
      <td>84.64%</td>
      <td>-0.13%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>295</td>
      <td>292</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>53</td>
      <td>53</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/solid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/solid_properties)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.64%</td>
      <td>90.64%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7905</td>
      <td>7902</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>816</td>
      <td>816</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/stochastic_tools)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.90%</td>
      <td>84.90%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26939</td>
      <td>26935</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4792</td>
      <td>4792</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/tensor_mechanics)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.11%</td>
      <td>89.11%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13866</td>
      <td>13863</td>
      <td>-3</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1694</td>
      <td>1694</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/thermal_hydraulics)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>cfd4d8</th>
      <th colspan=4>#25994 6e5a77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.29%</td>
      <td>82.28%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6839</td>
      <td>6836</td>
      <td>-3</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1472</td>
      <td>1472</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25994/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25994/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.26% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1947473/) on 6e5a773 wanted to post the following:


The following coverage requirement(s) failed:
- `misc` coverage rate 32.37% is less than the required 33.0%

MengnanLi91 committed:
Address review: rename functions, remove unnecesscary friend class

Co-Authored-By: brandonlangley <21372118+brandonlangley@users.noreply.github.com>

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1948428/) on e8bb7f9 wanted to post the following:


The following coverage requirement(s) failed:
- `misc` coverage rate 32.37% is less than the required 33.0%

loganharbour committed:
Define perf_log in Moose.C

loganharbour committed:
Clean up the use of root in MooseServer

loganharbour committed:
Remove perf_log instantiations

loganharbour committed:
Remove use of PerfLog

loganharbour committed:
Undo formatting change

loganharbour committed:
Remove extraneous newlines

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1955266/) on c007880 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7d1e620dc3e25b783ffb8251ccb6fae15127be4f`

loganharbour committed:
Remove another extraneous newline

loganharbour committed:
Remove timing from object that does not have a PerfGraph context

loganharbour committed:
Remove the defintion in Moose.C

loganharbour commented:
@brandonlangley FYI, cleaned up `MooseServer` some in https://github.com/idaholab/moose/pull/25994/commits/7e22b7ebd83b92708be7a44569b371f70001891a

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1955802/) on f83a2f1 wanted to post the following:


The following coverage requirement(s) failed:
- `misc` coverage rate 32.37% is less than the required 33.0%

loganharbour reviewed (changes requested):
Few small changes, did a full review. Should be good to go after these and the app fixes.

MengnanLi91 committed:
Address review: code and comment improvements

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1958225/) on 26f7323 wanted to post the following:


The following coverage requirement(s) failed:
- `misc` coverage rate 32.37% is less than the required 33.0%

MengnanLi91 committed:
Change misc coverage requirement

MengnanLi91 commented:
> Few small changes, did a full review. Should be good to go after these and the app fixes.

I've address the comments. App fixes should be the only thing that left

brandonlangley commented:
@loganharbour - Should this reference any actual issue number separate from a commit tagging its own PR number?

loganharbour commented:
## Preemptive patches

- [ ] Griffin https://github.inl.gov/ncrc/griffin/pull/1691

## Patches for after

- [ ] Griffin https://github.inl.gov/ncrc/griffin/pull/1692

loganharbour commented:
All patches are in. waiting for devel -> master.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1958601/) on f109f7e wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Test timings](https://civet.inl.gov/job/1982290/) on f109f7e wanted to post the following:

View timings [here](https://mooseframework.inl.gov/docs/PRs/25994/timing)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1958601/) on f109f7e wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)

loganharbour committed:
Hold on to the Parser as a shared_ptr to verify that it exists, remove unused method

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1984017/) on 9bbb59f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 28bf8f4f287b56d948c110e55ea724380f959b56`

loganharbour committed:
Remove extraneous ;

loganharbour committed:
Unify input filenames

loganharbour committed:
Account for full path in test

loganharbour committed:
Don't force realpaths

loganharbour committed:
Finalize path management and simplify Parser construction

loganharbour committed:
Fix an assertion

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1992637/) on 66c8a45 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25994/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8ac5e2a1afb934d8b46e1d1df9b945d5a0dc26c1`

loganharbour committed:
Clang format: how much I love and hate thee

loganharbour committed:
Revert back to MooseUtils::realpath

loganharbour committed:
Correct the input files logic

loganharbour committed:
Remove exposing the absolute paths outside of the Parser

loganharbour committed:
Still expose absolute paths to the Builder for FileName types

loganharbour committed:
Simplify root() call

loganharbour committed:
Make an error instead

permcody reviewed (approved):
Looks good Mengnan - nice work!
